pulsar-client/src/main/java/org/apache/pulsar/client/impl/TableViewImpl.java

```diff
@@ -232,33 +232,33 @@ private void handleMessage(Message<T> msg) {
 
     @Override
     public CompletableFuture<Void> refreshAsync() {
-        return reader.thenCompose(this::readAllExistingMessages);
+        return reader.thenCompose(reader -> readAllExistingMessages(reader, false));
```

Comment:
There is another concern. Since `readTailMessages` is running at the background, we don't have to call `readNextAsync` in another thread. Maybe we can just call `hasMessageAvailable` in a loop?

Reply:
Maybe we should consider having a callback queue for the refresh API. For example:

```
| LastMessageIds -> Future | LastMessageIds -> Future | LastMessageIds -> Future | LastMessageIds -> Future |
```

For the refresh request, we can try to get the last message IDs from the broker and then add it along with the Future to the queue.

```
| [{topic1-> (1,2), topic2 -> {3,2}}] -> Future |
```

Then returns the future to the caller.

The tailing read will check the header of the queue and try to complete the future if all the last message IDs are read.

Reply:
Good suggestion!